### PR TITLE
[SPARK-39825][SQL] Fix PushDownLeftSemiAntiJoin push through project

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

This pr improve the rule `PushDownLeftSemiAntiJoin` that forbid push left semi/anti through project by checking:
- probably pruned project
- complex join condition 

### Why are the changes needed?

Push LeftSemi/LeftAnti through project is not always effective

1. project is used to do column pruning, so it conflicts with ColumnPruning
2. if the project contains complex expression and join condition reference it, it can be regression since the complex expression will be evaluated more than once

### Does this PR introduce _any_ user-facing change?

yes

### How was this patch tested?

add test

The gloden file changes:
- q8: due to we forbid push through project if join condition reference a complex expression
- q10: due to we forbid push through project if it maybe a pruned project

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala [hunk 1]
```diff
@@ -1440,4 +1440,18 @@ class JoinSuite extends QueryTest with SharedSparkSession with AdaptiveSparkPlan
...
+    // before fix, it would throw:
+    //   java.lang.RuntimeException: Max iterations (100) reached for batch Operator Optimization
+    //   before Inferring Filters, please set 'spark.sql.optimizer.maxIterations' to a larger value
+    withTable("t") {
+      Seq((1, DoubleData(1, "a"))).toDF("c", "nested")
+        .write
+        .saveAsTable("t")
+      spark.sql("select c, nested.id from t")
+        .join(Seq(1).toDF("c"), Seq("c"), "left_semi")
+        .collect()
```
Comment (ulysses-you) [inline sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala:1454]: For a plan:
```sql
Project [i#1314, v#1572]
+- Join LeftSemi, (i#1314 = i#1579)
   :- Project [i#1314, data#1315.v AS v#1572]
   :  +- Relation spark_catalog.default.tbl[i#1314,data#1315,v#1316] parquet
   +- Project [value#1576 AS i#1579]
      +- LocalRelation [value#1576]
```

The Optimizer workflow:
1. PushDownLeftSemiAntiJoin push join through projet
2. ColumnPruning NestedColumnAliasing generate a new alias for nested column and add a project below join
3. CollapseProject and RemoveNoopOperators clean up the unnecessary project, so the plan is going to the beginning 

This issue also exists for non-nested column, but the batch is marked as not effective, the issue is not getting worse.
For nested column pruning we will create a new alias, so after applying the whole batch the plan is different with previous due to the different alias expr id.

**Comment by ulysses-you** [comment]
cc @cloud-fan @viirya thank you

**Comment by cloud-fan** [comment]
cc @sigmod

**Comment by cloud-fan** [comment]
BTW, does filter pushdown have the same issue?

**Comment by ulysses-you** [comment]
> BTW, does filter pushdown have the same issue?

no

The conflicts issue has been hacky fixed in ColumnPruning
https://github.com/apache/spark/blob/c2536a7eabd8764cbbaaff22935e19685b92f22b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala#L819-L829

Due to the subexpression elimination in codegen, the complex expression is not an issue too.

**Pull request closed without merging**